csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Dialogs;
+using Microsoft.Extensions.Options;
+
+namespace MessageRoutingBot
+{
+    /// <summary>
+    /// Main entry point and orchestration for bot.
+    /// </summary>
+    public class MessageRoutingBot : IBot
+    {
+        private readonly BotServices _services;
+        private readonly SemaphoreSlim _semaphore;
+        private readonly MessageRoutingBotAccessors _accessors;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EnterpriseBot"/> class.
+        /// </summary>
+        /// <param name="botServices">Bot services.</param>
+        /// <param name="accessors">Bot State Accessors.</param>
+        public MessageRoutingBot(BotServices botServices, MessageRoutingBotAccessors accessors)
+        {
+            _accessors = accessors;
+            _services = botServices;
+
+            // a semaphore to serialize access to the bot state
+            _semaphore = accessors.SemaphoreSlim;
+
+            Dialogs = new DialogSet(accessors.ConversationDialogState);
+            Dialogs.Add(new MainDialog(_services));
+        }
+
+        private DialogSet Dialogs { get; set; }
+
+        /// <summary>
+        /// Run every turn of the conversation. Handles orchestration of messages.
+        /// </summary>
+        /// <param name="turnContext">Bot Turn Context.</param>
+        /// <param name="cancellationToken">Task CancellationToken.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken)
+        {
+            try
+            {
+                await _semaphore.WaitAsync();
+
```

Comment:
It's far from obvious what this is doing. I'm also not convinced it's right. 

We want to (if I understand correctly) serialize on a per-conversation basis. Got globally serialized.

Reply:
We will remove this. When this was started, this was one of the patterns we were considering for the enterprise bot but not anymore!